diff --git a/CMakeLists.txt b/CMakeLists.txt
index 86e0affc4..d6d328f50 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -584,6 +584,13 @@ include_directories(${CMAKE_CURRENT_SOURCE_DIR} ${ZLIB_INCLUDE_DIRS})
 
 unset(PNG_LIB_TARGETS)
 
+if(EMSCRIPTEN)
+   message("Detected emscripten and enabling pthreads")
+   set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -pthread")
+   set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -pthread")
+   set(CMAKE_LINKER_FLAGS "${CMAKE_LINKER_FLAGS} -pthread")
+endif()
+
 if(PNG_SHARED)
   add_library(png SHARED ${libpng_sources})
   set(PNG_LIB_TARGETS png)
